### ISSUE TITLE
input specification in app metadata

### ISSUE THREAD

**Issue by keighrim** [clamsproject/clams-python#50]
This thread to discuss the design of output of apps as specified in their `appmetadata`, particularly in the `requires` field.

**Comment by keighrim** [comment]
From a new development on the versioning issue (https://github.com/clamsproject/mmif/issues/14, https://github.com/clamsproject/mmif-python/issues/163), now `input` field cannot simply list version-specific `@type`s to accurately express backward (and possibly forward) compatibility with other versions of the same `@type`s. 
Here are some ideas; 

1. use a wild card character to indicate the compatibility.  (`*`, `?`, `x`, `X`, etc.)
`http://.../1.0.*/TimeFrame` is compatible with anything from `.../1.0.0/TimeFrame` to `.../1.0.999⋯`. 
1. omit the patch (or patch + minor) number from its string representation. 
`http://.../1.0/TimeFrame` is compatible with anything from `.../1.0.0/TimeFrame` to `.../1.0.999⋯`. 
1. just leave the full version in the string but document that `@type`s in the `input` list must be considered with their compatible other version (not a good solution, I think)

**Comment by marcverhagen** [comment]
I am not sure how much of an issue this should be. In the input I usually do something like

```
[{'@type': AnnotationTypes.Timeframe.value}]
``` 

Which means that any version comes from a default inside of the mmif-sdk, basically the current version. I am not wild about having to say

```
[{'@type': 'http:/mmif.clasm.ai/VERSION_SPEC/Timeframe'}]
``` 

Using a wildcard in there or part of the version makes this a non-existing URL, which I don't like. An app assumes a particular version of MMIF, and should just also assume that its input uses that version of MMIF. The platform and the pipeline should take care of anything outside of that in case an upstream component uses an older version of MMIF:

- Differences in major versions are always breaking.
- MMIF patch versions don't matter, so the pipeline should not break
- Differences in minor version may matter. The app could check whether the input provides what it needs using the input specifications, so if it requires a boxType=text and there is no input like that it could raise a warning and not apply. We have a data driven platform and not a pipeline platform. If the data for some app ain't there, it won't do anything.

What I am saying is that we should not put stuff in the input that does not fit there (like version ranges), but experiment with what happens when we build a pipeline with out-of-date upstream tools or run an app over old MMIF.

Of the three options above this is closest to 3 I am afraid.

**Comment by keighrim** [comment]
I got your point. And I fully agree that having non-existing URLs shown in the MMIF is not good. Previously when I said, 
> just leave the full version in the string but document that @types in the input list must be considered with their compatible other version (not a good solution, I think)

the main reason I thought it wasn't a good solution was because of lack of transparency when you put the full URL there. For example, when an app says it can take a input of `http://.../1.0.3/vocab/TimeFrame`, it can actually take inputs like `http://.../1.0.0/vocab/TimeFrame`, `http://.../1.0.2/vocab/TimeFrame`, or maybe even `http://.../1.0.13/vocab/TimeFrame`. 

I wasn't thinking clearly on the issue with non-existing URLs at the time, and now I believe having full version specified in the `input` list isn't that bad solution compared to the others.

**Comment by marcverhagen** [comment]
Yeah, potential lack of transparancy is indeed an issue. But it is somewhat hidden from the user because they do not put in a full URL but just a type from the annotations type list.

**Comment by keighrim** [comment]
closed via #53 .
